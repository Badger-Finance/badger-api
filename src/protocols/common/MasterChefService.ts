import { Service } from "@tsed/di";
import NodeCache from "node-cache";
import { MasterChefPool } from "../../interface/MasterChef";
import { Chain } from "../../util/constants";
import { getProvider } from "../../util/util";

@Service()
export class MasterChefService {
  private cache: NodeCache;

  constructor() {
    this.cache = new NodeCache({ stdTTL: 300, checkperiod: 480 });
  }

  async getPool(chain: Chain, chef: string, poolId: number): Promise<MasterChefPool> {
    const provider = getProvider(chain);
  }
}
