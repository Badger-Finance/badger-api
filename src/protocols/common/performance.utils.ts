import fetch from 'node-fetch';
import { Performance, uniformPerformance } from '../interfaces/performance.interface';
import { createValueSource, ValueSource } from '../interfaces/value-source.interface';

export async function getSwapValueSource(graphUrl: string, name: string, poolAddress: string): Promise<ValueSource> {
  // TODO: Move query to GraphService
  const query = `
    {
      pairDayDatas(first: 30, orderBy: date, orderDirection: desc, where:{pairAddress: "${poolAddress.toLowerCase()}"}) {
        reserveUSD
        dailyVolumeUSD
      }
    }
  `;
  const response = await fetch(graphUrl, {
    method: 'POST',
    body: JSON.stringify({ query }),
  });

  const sourceName = `${name} LP Fees`;
  const unknownValueSource = createValueSource(sourceName, uniformPerformance(0));

  if (!response.ok) {
    return unknownValueSource;
  }

  const pairDayResponse = await response.json();
  if (pairDayResponse.errors) {
    return unknownValueSource;
  }
  const pairDayData = pairDayResponse.data.pairDayDatas;

  let totalApy = 0;
  const performance: Performance = uniformPerformance(0);
  for (let i = 0; i < pairDayData.length; i++) {
    const volume = parseFloat(pairDayData[i].dailyVolumeUSD);
    const poolReserve = parseFloat(pairDayData[i].reserveUSD);
    const fees = volume * 0.003;
    totalApy += (fees / poolReserve) * 365 * 100;
    const currentApy = totalApy / (i + 1);
    if (i === 0) performance.oneDay = currentApy;
    if (i === 2) performance.threeDay = currentApy;
    if (i === 6) performance.sevenDay = currentApy;
    if (i === 29) performance.thirtyDay = currentApy;
  }

  return createValueSource(sourceName, performance);
}
