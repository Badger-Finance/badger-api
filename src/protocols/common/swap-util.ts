import { NotFound, UnprocessableEntity } from '@tsed/exceptions';
import { BigNumber, ethers } from 'ethers';
import { GraphQLClient } from 'graphql-request';
import { Chain } from '../../chains/config/chain.config';
import { Network } from '../../chains/enums/chain-network.enum';
import { uniV2LPAbi } from '../../config/abi';
import { SUSHISWAP_URL, UNISWAP_URL } from '../../config/constants';
import { getSdk as getUniswapSdk } from '../../graphql/generated/uniswap';
import { getTokenPriceData } from '../../prices/prices-util';
import { TokenPrice } from '../../tokens/interfaces/token-price.interface';

interface LiquidityData {
  contract: string;
  token0: string;
  token1: string;
  reserve0: number;
  reserve1: number;
  totalSupply: number;
}

interface GetReservesResponse {
  _reserve0: BigNumber;
  _reserve1: BigNumber;
  _blockTimestampLast: number;
}

export const getLiquidityData = async (chain: Chain, contract: string): Promise<LiquidityData> => {
  const pairContract = new ethers.Contract(contract, uniV2LPAbi, chain.provider);
  const totalSupply = parseFloat(ethers.utils.formatEther(await pairContract.totalSupply()));
  const token0 = await pairContract.token0();
  const token1 = await pairContract.token1();
  const reserves: GetReservesResponse = await pairContract.getReserves();
  const reserve0 = parseFloat(ethers.utils.formatEther(reserves._reserve0));
  const reserve1 = parseFloat(ethers.utils.formatEther(reserves._reserve1));
  return {
    contract: contract,
    token0: token0,
    token1: token1,
    reserve0: reserve0,
    reserve1: reserve1,
    totalSupply: totalSupply,
  };
};

export const getLiquidityPrice = async (graphUrl: string, contract: string): Promise<TokenPrice> => {
  const graphqlClient = new GraphQLClient(graphUrl);
  const graphqlSdk = getUniswapSdk(graphqlClient);
  const { pair } = await graphqlSdk.UniV2Pair({
    id: contract.toLowerCase(),
  });
  if (!pair) {
    throw new NotFound(`No pair found for ${contract}`);
  }
  if (pair.totalSupply === 0) {
    return {
      address: contract,
      usd: 0,
      eth: 0,
    };
  }
  const liquidityData: LiquidityData = {
    contract: contract,
    token0: pair.token0.id,
    token1: pair.token1.id,
    reserve0: pair.reserve0,
    reserve1: pair.reserve1,
    totalSupply: pair.totalSupply,
  };
  return resolveLiquidityPrice(liquidityData);
};

export const getOnChainLiquidityPrice = async (chain: Chain, contract: string): Promise<TokenPrice> => {
  try {
    const liquidityData = await getLiquidityData(chain, contract);
    if (liquidityData.totalSupply === 0) {
      return {
        address: contract,
        usd: 0,
        eth: 0,
      };
    }
    return resolveLiquidityPrice(liquidityData);
  } catch (err) {
    throw new NotFound(`No pair found for ${contract}`);
  }
};

const resolveLiquidityPrice = async (liquidityData: LiquidityData): Promise<TokenPrice> => {
  const { contract, token0, token1, reserve0, reserve1, totalSupply } = liquidityData;
  let t0Price = await getTokenPriceData(token0);
  let t1Price = await getTokenPriceData(token1);
  if (!t0Price && !t1Price) throw new UnprocessableEntity(`Token pair ${contract} cannot be priced`);
  if (!t0Price) {
    const t1Scalar = reserve0 / reserve1;
    t0Price = {
      address: token0,
      usd: t1Price.usd * t1Scalar,
      eth: t1Price.eth * t1Scalar,
    };
  }
  if (!t1Price) {
    const t0Scalar = reserve1 / reserve0;
    t1Price = {
      address: token1,
      usd: t0Price.usd * t0Scalar,
      eth: t0Price.eth * t0Scalar,
    };
  }
  const usdPrice = (t0Price.usd * reserve0 + t1Price.usd * reserve1) / totalSupply;
  const ethPrice = (t0Price.eth * reserve0 + t1Price.eth * reserve1) / totalSupply;
  return {
    address: contract,
    usd: usdPrice,
    eth: ethPrice,
  };
};

export const getSushiswapPrice = async (contract: string): Promise<TokenPrice> => {
  return getLiquidityPrice(SUSHISWAP_URL, contract);
};

export const getUniswapPrice = async (contract: string): Promise<TokenPrice> => {
  return getLiquidityPrice(UNISWAP_URL, contract);
};

export const getPancakeswapPrice = async (contract: string): Promise<TokenPrice> => {
  return getOnChainLiquidityPrice(Chain.getChain(Network.BinanceSmartChain), contract);
};
