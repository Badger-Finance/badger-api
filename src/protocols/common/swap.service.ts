import { Chain } from '../../chains/config/chain.config';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { TokenPrice } from '../../tokens/interfaces/token-price.interface';
import { ValueSource } from '../interfaces/value-source.interface';
import { getSwapValueSource } from './performance.utils';
import { getLiquidityPrice } from './swap.utils';

export abstract class SwapService {
  constructor(private graphUrl: string, private name: string) {}

  abstract getPairPerformance(
    chain: Chain,
    sett: SettDefinition,
    filterHarvestablePerformances?: boolean,
  ): Promise<ValueSource[]>;

  /**
   * Retrieve Uniswap v2 variant pool performance from trading fees.
   * @param poolAddress Liquidity pair contract address.
   * @param protocol Uniswap v2 variant type.
   */
  async getSwapPerformance(poolAddress: string): Promise<ValueSource> {
    return getSwapValueSource(this.graphUrl, this.name, poolAddress);
  }

  async getPairPrice(contract: string): Promise<TokenPrice> {
    return getLiquidityPrice(this.graphUrl, contract);
  }
}
