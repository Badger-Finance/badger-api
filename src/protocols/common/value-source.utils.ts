import { CachedValueSource } from '../interfaces/cached-value-source.interface';
import { ValueSource } from '../interfaces/value-source.interface';

export function cachedValueSourceToValueSource(cachedValueSource: CachedValueSource): ValueSource {
  return {
    name: cachedValueSource.name,
    apy: cachedValueSource.apy,
    harvestable: cachedValueSource.harvestable,
    performance: {
      oneDay: cachedValueSource.oneDay,
      threeDay: cachedValueSource.threeDay,
      sevenDay: cachedValueSource.sevenDay,
      thirtyDay: cachedValueSource.thirtyDay,
    },
  };
}
