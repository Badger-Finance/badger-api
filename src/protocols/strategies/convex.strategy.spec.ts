import {
  CurveBaseRegistry,
  CurveBaseRegistry__factory,
  CurveRegistry,
  CurveRegistry__factory,
  Erc20,
  Erc20__factory,
  Network,
  TokenValue,
} from '@badger-dao/sdk';
import { BigNumber, BigNumberish, ethers } from 'ethers';
import { mock, MockProxy } from 'jest-mock-extended';

import { Vaultish } from '../../aws/interfaces/vaultish.interface';
import { CurrentVaultSnapshotModel } from '../../aws/models/current-vault-snapshot.model';
import { VaultDefinitionModel } from '../../aws/models/vault-definition.model';
import { Chain } from '../../chains/config/chain.config';
import * as requestUtils from '../../common/request';
import { TOKENS } from '../../config/tokens.config';
import { CurvePool, CurvePool__factory, CurvePool3, CurvePool3__factory } from '../../contracts';
import { PromiseOrValue } from '../../contracts/common';
import { MOCK_TOKENS, MOCK_VAULT_DEFINITION, MOCK_VAULT_SNAPSHOT, TEST_ADDR, TEST_TOKEN } from '../../test/constants';
import { mockBalance, setupMockChain } from '../../test/mocks.utils';
import { setupTestVault } from '../../test/mocks.utils/mock.api.chain';
import * as tokensUtils from '../../tokens/tokens.utils';
import * as vaultsUtils from '../../vaults/vaults.utils';
import {
  getCurvePerformance,
  getCurvePoolBalance,
  getCurveTokenPrice,
  getCurveVaultTokenBalance,
  getCurveYieldSources,
  resolveCurvePoolTokenPrice,
  resolveCurveStablePoolTokenPrice,
} from './convex.strategy';

describe('convex.strategy', () => {
  let chain: Chain;
  let baseRegistry: MockProxy<CurveBaseRegistry>;
  let registry: MockProxy<CurveRegistry>;
  let curvePool3: MockProxy<CurvePool3>;
  let curvePool: MockProxy<CurvePool>;

  function setupBadPoolSize() {
    async function bad_coins(index: PromiseOrValue<BigNumberish>): Promise<string> {
      switch (index) {
        case 0:
          return TOKENS.WBTC;
        case 1:
          return TOKENS.DIGG;
        case 2:
          return TOKENS.BADGER;
        default:
          throw new Error('Expected test error: coins');
      }
    }
    jest.spyOn(curvePool3, 'coins').mockImplementation(bad_coins);
    jest.spyOn(curvePool, 'coins').mockImplementation(bad_coins);

    async function bad_balances(index: PromiseOrValue<BigNumberish>): Promise<BigNumber> {
      switch (index) {
        case 0:
          return ethers.constants.WeiPerEther;
        case 1:
          return ethers.constants.WeiPerEther;
        case 2:
          return ethers.constants.WeiPerEther.mul(5000);
        default:
          throw new Error('Expected test error: balances');
      }
    }
    jest.spyOn(curvePool3, 'balances').mockImplementation(bad_balances);
    jest.spyOn(curvePool, 'balances').mockImplementation(bad_balances);
  }

  beforeEach(() => {
    chain = setupMockChain();
    baseRegistry = mock<CurveBaseRegistry>();
    registry = mock<CurveRegistry>();
    curvePool3 = mock<CurvePool3>();
    curvePool = mock<CurvePool>();

    jest.spyOn(baseRegistry, 'get_registry').mockImplementation(async () => TEST_ADDR);
    jest.spyOn(CurveBaseRegistry__factory, 'connect').mockImplementation(() => baseRegistry);

    jest.spyOn(registry, 'get_pool_from_lp_token').mockImplementation(async () => TEST_ADDR);
    jest.spyOn(CurveRegistry__factory, 'connect').mockImplementation(() => registry);

    async function coins(index: PromiseOrValue<BigNumberish>): Promise<string> {
      switch (index) {
        case 0:
          return TOKENS.WBTC;
        case 1:
          return TOKENS.BADGER;
        default:
          throw new Error('Expected test error: coins');
      }
    }
    jest.spyOn(curvePool3, 'coins').mockImplementation(coins);
    jest.spyOn(curvePool, 'coins').mockImplementation(coins);

    async function balances(index: PromiseOrValue<BigNumberish>): Promise<BigNumber> {
      switch (index) {
        case 0:
          return ethers.constants.WeiPerEther;
        case 1:
          return ethers.constants.WeiPerEther.mul(5000);
        default:
          throw new Error('Expected test error: balances');
      }
    }
    jest.spyOn(curvePool3, 'balances').mockImplementation(balances);
    jest.spyOn(curvePool, 'balances').mockImplementation(balances);

    jest.spyOn(curvePool3, 'price_oracle').mockImplementation(async () => BigNumber.from('963845856610322279'));

    jest.spyOn(CurvePool3__factory, 'connect').mockImplementation(() => curvePool3);
    jest.spyOn(CurvePool__factory, 'connect').mockImplementation(() => curvePool);

    const erc20 = mock<Erc20>();
    jest.spyOn(erc20, 'totalSupply').mockImplementation(async () => ethers.constants.WeiPerEther);
    jest.spyOn(Erc20__factory, 'connect').mockImplementation(() => erc20);

    jest
      .spyOn(vaultsUtils, 'getCachedVault')
      .mockImplementation(async (_c, _v) => MOCK_VAULT_SNAPSHOT as CurrentVaultSnapshotModel);

    jest.spyOn(requestUtils, 'request').mockImplementation(async () => ({
      apy: {
        month: {
          tricrypto2: 0.1,
        },
        day: {
          tricrypto2: 0.04,
        },
        week: {
          tricrypto2: 0.2,
        },
      },
    }));

    jest.spyOn(tokensUtils, 'getFullToken').mockImplementation(async (_, token) => MOCK_TOKENS[token]);
  });

  describe('getCurveYieldSources', () => {
    it('returns no sources for bvecvx', async () => {
      const bveCVXDefinition = JSON.parse(JSON.stringify(MOCK_VAULT_DEFINITION));
      bveCVXDefinition.address = TOKENS.BVECVX;
      const result = await getCurveYieldSources(chain, bveCVXDefinition);
      expect(result).toMatchObject([]);
    });

    it('returns trade fees for standard vaults', async () => {
      const tricryptoDefinition = JSON.parse(JSON.stringify(MOCK_VAULT_DEFINITION));
      tricryptoDefinition.depositToken = TOKENS.CRV_TRICRYPTO2;
      const result = await getCurveYieldSources(chain, tricryptoDefinition);
      expect(result).toMatchSnapshot();
    });

    it('returns derived data for liquidity pool vault', async () => {
      setupTestVault();

      jest
        .spyOn(tokensUtils, 'getVaultTokens')
        .mockImplementation(async (_chain: Chain, _vault: Vaultish, _currency?: string): Promise<TokenValue[]> => {
          const token0 = MOCK_TOKENS[TOKENS.CVX];
          const token1 = MOCK_TOKENS[TOKENS.BVECVX];
          return [mockBalance(token0, 15), mockBalance(token1, 16)];
        });
      let calls = 0;
      jest.spyOn(requestUtils, 'request').mockImplementation(async () => {
        if (calls <= 1) {
          calls++;
          return {
            apy: { week: {} },
          };
        }
        return {
          data: {
            poolDetails: [
              {
                apy: 1.2,
                poolAddress: TOKENS.BCRV_CVXBVECVX,
              },
            ],
          },
        };
      });
      const liquidityDefinition = JSON.parse(JSON.stringify(MOCK_VAULT_DEFINITION));
      liquidityDefinition.address = TOKENS.BCRV_CVXBVECVX;
      const result = await getCurveYieldSources(chain, liquidityDefinition);
      expect(result).toMatchSnapshot();
    });
  });

  describe('getCurvePoolBalance', () => {
    it('returns curve pool underlying pool', async () => {
      const result = await getCurvePoolBalance(chain, TEST_TOKEN);
      expect(result).toMatchSnapshot();
    });

    it('falls back to pool definitions for pools missing from the registry', async () => {
      jest.spyOn(registry, 'get_pool_from_lp_token').mockImplementation(async () => ethers.constants.AddressZero);
      const result = await getCurvePoolBalance(chain, TOKENS.CVXFXS);
      expect(result).toMatchSnapshot();
    });

    it('falls back to older pool abi when new pools fail', async () => {
      jest.spyOn(curvePool3, 'coins').mockImplementation(async () => {
        throw new Error('Expected test error: coins');
      });
      const result = await getCurvePoolBalance(chain, TEST_TOKEN);
      expect(result).toMatchSnapshot();
    });

    it('returns no tokens if all abi fail', async () => {
      jest.spyOn(curvePool3, 'coins').mockImplementation(async () => {
        throw new Error('Expected test error: coins');
      });
      jest.spyOn(curvePool, 'coins').mockImplementation(async () => {
        throw new Error('Expected test error: coins');
      });
      const result = await getCurvePoolBalance(chain, TEST_TOKEN);
      expect(result).toMatchObject([]);
    });
  });

  describe('getCurveVaultTokenBalance', () => {
    it('returns vault tokens built on a curve pool', async () => {
      const result = await getCurveVaultTokenBalance(chain, MOCK_VAULT_DEFINITION);
      expect(result).toMatchSnapshot();
    });
  });

  describe('getCurveTokenPrice', () => {
    it('returns the price of a curve pool token', async () => {
      const result = await getCurveTokenPrice(chain, TEST_TOKEN);
      expect(result).toMatchSnapshot();
    });
  });

  describe('resolveCurvePoolTokenPrice', () => {
    it('throws an error for wrong pool size', async () => {
      setupBadPoolSize();
      await expect(resolveCurvePoolTokenPrice(chain, MOCK_TOKENS[TOKENS.DIGG])).rejects.toThrow(
        'Pool has unexpected number of tokens!',
      );
    });

    it('resolves the requested token0 price based off paired token price', async () => {
      const result = await resolveCurvePoolTokenPrice(chain, MOCK_TOKENS[TOKENS.WBTC]);
      expect(result).toMatchSnapshot();
    });

    it('resolves the requested token1 price based off paired token price', async () => {
      const result = await resolveCurvePoolTokenPrice(chain, MOCK_TOKENS[TOKENS.BADGER]);
      expect(result).toMatchSnapshot();
    });
  });

  describe('resolveCurveStablePoolTokenPrice', () => {
    it('throws an error for wrong pool size', async () => {
      setupBadPoolSize();
      await expect(resolveCurveStablePoolTokenPrice(chain, MOCK_TOKENS[TOKENS.CVXFXS])).rejects.toThrow(
        'Pool has unexpected number of tokens!',
      );
    });

    it('returns price derived from pool price oracle', async () => {
      const result = await resolveCurveStablePoolTokenPrice(chain, MOCK_TOKENS[TOKENS.CVXFXS]);
      expect(result).toMatchSnapshot();
    });

    it('returns price of zero when encounters pool price oracle error', async () => {
      jest.spyOn(console, 'error').mockImplementation();
      jest.spyOn(curvePool3, 'price_oracle').mockImplementation(async () => {
        throw new Error('Expected test error: price_oracle');
      });
      const { price } = await resolveCurveStablePoolTokenPrice(chain, MOCK_TOKENS[TOKENS.CVXFXS]);
      expect(price).toEqual(0);
    });
  });

  describe('getCurvePerformance', () => {
    let tricryptoDefinition: VaultDefinitionModel;

    beforeEach(() => {
      tricryptoDefinition = JSON.parse(JSON.stringify(MOCK_VAULT_DEFINITION));
      tricryptoDefinition.depositToken = TOKENS.CRV_TRICRYPTO2;
    });

    it.each([[Network.Arbitrum], [Network.Polygon], [Network.Ethereum]])(
      'requests trade fees from the curve api on %s',
      async (network) => {
        chain = setupMockChain({ network });
        const result = await getCurvePerformance(chain, tricryptoDefinition);
        expect(result).toMatchSnapshot();
      },
    );

    it('utilizes crypto api fallback option if required', async () => {
      let calls = 0;
      jest.spyOn(requestUtils, 'request').mockImplementation(async () => {
        if (calls === 0) {
          calls++;
          return {
            apy: { week: {} },
          };
        }
        return {
          apy: {
            month: {
              tricrypto2: 0.01,
            },
            day: {
              tricrypto2: 0.4,
            },
            week: {
              tricrypto2: 0.02,
            },
          },
        };
      });
    });

    it('utilizes factory 2 api fallback option if required', async () => {
      let calls = 0;
      jest.spyOn(requestUtils, 'request').mockImplementation(async () => {
        if (calls <= 1) {
          calls++;
          return {
            apy: { week: {} },
          };
        }
        return {
          data: {
            poolDetails: [
              {
                apy: 1.2,
                poolAddress: TOKENS.CRV_TRICRYPTO2,
              },
            ],
          },
        };
      });
      const result = await getCurvePerformance(chain, tricryptoDefinition);
      expect(result).toMatchSnapshot();
    });

    it('utilizes crypto factory api fallback option if required', async () => {
      let calls = 0;
      jest.spyOn(requestUtils, 'request').mockImplementation(async () => {
        if (calls <= 1) {
          calls++;
          return {
            apy: { week: {} },
          };
        }
        if (calls <= 2) {
          calls++;
          return {
            data: {
              poolDetails: [],
            },
          };
        }
        return {
          data: {
            poolDetails: [
              {
                apy: 1.2,
                poolAddress: TOKENS.CRV_TRICRYPTO2,
              },
            ],
          },
        };
      });
      const result = await getCurvePerformance(chain, tricryptoDefinition);
      expect(result).toMatchSnapshot();
    });
  });
});
