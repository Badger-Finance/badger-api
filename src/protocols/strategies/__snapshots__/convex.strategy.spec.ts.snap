// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convex.strategy getCurvePerformance requests trade fees from the curve api on arbitrum 1`] = `
CachedYieldSource {
  "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
  "boostable": false,
  "chain": "ethereum",
  "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
  "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_curve_lp_fees_flat",
  "name": "Curve LP Fees",
  "performance": Object {
    "baseYield": 20,
    "grossYield": 20,
    "maxGrossYield": 20,
    "maxYield": 20,
    "minGrossYield": 20,
    "minYield": 20,
  },
  "type": "lp_fee",
}
`;

exports[`convex.strategy getCurvePerformance requests trade fees from the curve api on ethereum 1`] = `
CachedYieldSource {
  "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
  "boostable": false,
  "chain": "ethereum",
  "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
  "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_curve_lp_fees_flat",
  "name": "Curve LP Fees",
  "performance": Object {
    "baseYield": 20,
    "grossYield": 20,
    "maxGrossYield": 20,
    "maxYield": 20,
    "minGrossYield": 20,
    "minYield": 20,
  },
  "type": "lp_fee",
}
`;

exports[`convex.strategy getCurvePerformance requests trade fees from the curve api on polygon 1`] = `
CachedYieldSource {
  "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
  "boostable": false,
  "chain": "ethereum",
  "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
  "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_curve_lp_fees_flat",
  "name": "Curve LP Fees",
  "performance": Object {
    "baseYield": 20,
    "grossYield": 20,
    "maxGrossYield": 20,
    "maxYield": 20,
    "minGrossYield": 20,
    "minYield": 20,
  },
  "type": "lp_fee",
}
`;

exports[`convex.strategy getCurvePerformance utilizes crypto factory api fallback option if required 1`] = `
CachedYieldSource {
  "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
  "boostable": false,
  "chain": "ethereum",
  "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
  "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_curve_lp_fees_flat",
  "name": "Curve LP Fees",
  "performance": Object {
    "baseYield": 1.2,
    "grossYield": 1.2,
    "maxGrossYield": 1.2,
    "maxYield": 1.2,
    "minGrossYield": 1.2,
    "minYield": 1.2,
  },
  "type": "lp_fee",
}
`;

exports[`convex.strategy getCurvePerformance utilizes factory 2 api fallback option if required 1`] = `
CachedYieldSource {
  "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
  "boostable": false,
  "chain": "ethereum",
  "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
  "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_curve_lp_fees_flat",
  "name": "Curve LP Fees",
  "performance": Object {
    "baseYield": 1.2,
    "grossYield": 1.2,
    "maxGrossYield": 1.2,
    "maxYield": 1.2,
    "minGrossYield": 1.2,
    "minYield": 1.2,
  },
  "type": "lp_fee",
}
`;

exports[`convex.strategy getCurvePoolBalance falls back to older pool abi when new pools fail 1`] = `
Array [
  Object {
    "address": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
    "balance": 10000000000,
    "decimals": 8,
    "name": "Wrapped BTC",
    "symbol": "WBTC",
    "value": 88000000000000,
  },
  Object {
    "address": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
    "balance": 5000,
    "decimals": 18,
    "name": "Badger",
    "symbol": "BADGER",
    "value": 67130000,
  },
]
`;

exports[`convex.strategy getCurvePoolBalance falls back to pool definitions for pools missing from the registry 1`] = `
Array [
  Object {
    "address": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
    "balance": 10000000000,
    "decimals": 8,
    "name": "Wrapped BTC",
    "symbol": "WBTC",
    "value": 88000000000000,
  },
  Object {
    "address": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
    "balance": 5000,
    "decimals": 18,
    "name": "Badger",
    "symbol": "BADGER",
    "value": 67130000,
  },
]
`;

exports[`convex.strategy getCurvePoolBalance returns curve pool underlying pool 1`] = `
Array [
  Object {
    "address": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
    "balance": 10000000000,
    "decimals": 8,
    "name": "Wrapped BTC",
    "symbol": "WBTC",
    "value": 88000000000000,
  },
  Object {
    "address": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
    "balance": 5000,
    "decimals": 18,
    "name": "Badger",
    "symbol": "BADGER",
    "value": 67130000,
  },
]
`;

exports[`convex.strategy getCurveTokenPrice returns the price of a curve pool token 1`] = `
Object {
  "address": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
  "price": 88000067130000,
}
`;

exports[`convex.strategy getCurveVaultTokenBalance returns vault tokens built on a curve pool 1`] = `
VaultTokenBalance {
  "chain": "ethereum",
  "id": "ethereum-0x37d9D2C6035b744849C15F1BFEE8F268a20fCBd8",
  "tokenBalances": Array [
    Object {
      "address": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
      "balance": 484054534550573.25,
      "decimals": 8,
      "name": "Wrapped BTC",
      "symbol": "WBTC",
      "value": 4259679904045044000,
    },
    Object {
      "address": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
      "balance": 242027267.2752866,
      "decimals": 18,
      "name": "Badger",
      "symbol": "BADGER",
      "value": 3249458090437.998,
    },
  ],
  "vault": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
}
`;

exports[`convex.strategy getCurveYieldSources returns derived data for liquidity pool vault 1`] = `
Array [
  CachedYieldSource {
    "address": "0x937B8E917d0F36eDEBBA8E459C5FB16F3b315551",
    "boostable": false,
    "chain": "ethereum",
    "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_curve_lp_fees_flat",
    "name": "Curve LP Fees",
    "performance": Object {
      "baseYield": 0,
      "grossYield": 0,
      "maxGrossYield": 0,
      "maxYield": 0,
      "minGrossYield": 0,
      "minYield": 0,
    },
    "type": "lp_fee",
  },
  CachedYieldSource {
    "address": "0x937B8E917d0F36eDEBBA8E459C5FB16F3b315551",
    "boostable": false,
    "chain": "ethereum",
    "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_compound_vault_compounding_flat",
    "name": "Vault Compounding",
    "performance": Object {
      "baseYield": 0.3196181652036132,
      "grossYield": 0.3196181652036132,
      "maxGrossYield": 0.3196181652036132,
      "maxYield": 0.3196181652036132,
      "minGrossYield": 0.3196181652036132,
      "minYield": 0.3196181652036132,
    },
    "type": "compound",
  },
  CachedYieldSource {
    "address": "0x937B8E917d0F36eDEBBA8E459C5FB16F3b315551",
    "boostable": false,
    "chain": "ethereum",
    "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_emission_badger_rewards_flat",
    "name": "Badger Rewards",
    "performance": Object {
      "baseYield": 0.3196181652036132,
      "grossYield": 0.3196181652036132,
      "maxGrossYield": 0.3196181652036132,
      "maxYield": 0.3196181652036132,
      "minGrossYield": 0.3196181652036132,
      "minYield": 0.3196181652036132,
    },
    "type": "emission",
  },
  CachedYieldSource {
    "address": "0x937B8E917d0F36eDEBBA8E459C5FB16F3b315551",
    "boostable": false,
    "chain": "ethereum",
    "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_curve_lp_fees_flat",
    "name": "Curve LP Fees",
    "performance": Object {
      "baseYield": 0.3196181652036132,
      "grossYield": 0.3196181652036132,
      "maxGrossYield": 0.3196181652036132,
      "maxYield": 0.3196181652036132,
      "minGrossYield": 0.3196181652036132,
      "minYield": 0.3196181652036132,
    },
    "type": "lp_fee",
  },
]
`;

exports[`convex.strategy getCurveYieldSources returns trade fees for standard vaults 1`] = `
Array [
  CachedYieldSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "boostable": false,
    "chain": "ethereum",
    "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_curve_lp_fees_flat",
    "name": "Curve LP Fees",
    "performance": Object {
      "baseYield": 20,
      "grossYield": 20,
      "maxGrossYield": 20,
      "maxYield": 20,
      "minGrossYield": 20,
      "minYield": 20,
    },
    "type": "lp_fee",
  },
]
`;

exports[`convex.strategy resolveCurvePoolTokenPrice resolves the requested token0 price based off paired token price 1`] = `
Object {
  "address": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
  "price": 0.006713,
}
`;

exports[`convex.strategy resolveCurvePoolTokenPrice resolves the requested token1 price based off paired token price 1`] = `
Object {
  "address": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
  "price": 17600000000,
}
`;

exports[`convex.strategy resolveCurveStablePoolTokenPrice returns price derived from pool price oracle 1`] = `
Object {
  "address": "0xFEEf77d3f69374f66429C91d732A244f074bdf74",
  "price": 8481.843538170837,
}
`;
